ffi_internals/src/type_ffi.rs

```diff
@@ -0,0 +1,534 @@
+            STRING | STR => Self::String,
+            UUID => Self::Uuid,
+            BOOL | U8 | U16 | U32 | U64 | I8 | I16 | I32 | I64 | F32 | F64 => Self::Raw(type_path),
+            _other => Self::Boxed(type_path),
+        }
+    }
+}
+
+/// The context in which a type is being referenced. Sometimes the mutability of a reference, or
+/// even the type it's exposed as is different depending on whether it's being returned to the
+/// consumer or passed in as an argument to a Rust function.
+///
+#[derive(Debug, Clone, Copy)]
+pub enum Context {
+    /// Type is being used as an argument to a Rust function.
+    Argument,
+    /// Type is being returned to the consumer.
+    Return,
+}
+
+/// Describes the supported language-level generic wrappers around a `NativeType`, so that we can
+/// expose an `Option<Foo>` or even a `Result<Vec<Foo>>`.
+///
+/// It's worth noting that these are only supported one level deep; we won't be able to expose a
+/// `Vec<Vec<Foo>>` without making some larger improvements to the way we parse types.
+///
+#[derive(Debug)]
+#[allow(clippy::struct_excessive_bools)]
+pub struct TypeFFI {
+    /// The underlying type being exposed.
+    ///
+    pub native_type: TypeIdentifier,
+    /// True if `native_type` is wrapped in an `Option`, otherwise false.
+    ///
+    pub is_option: bool,
+    /// True if `native_type` is the type of the elements in a `Vec` or slice, otherwise false.
+    ///
+    pub is_vec: bool,
+    /// True if `native_type` is the type of the `Success` variant of a `Result`, otherwise false.
+    ///
+    pub is_result: bool,
+    /// True if `native_type` is wrapped in a `Cow`, otherwise false.
+    ///
+    pub is_cow: bool,
+    /// True if we're dealing with a borrowed reference to `native_type`./
+    ///
+    pub is_borrow: bool,
+}
+
+impl From<(TypeIdentifier, WrappingType)> for TypeFFI {
+    fn from(data: (TypeIdentifier, WrappingType)) -> Self {
+        let (native_type, wrapping_type) = data;
+        Self {
+            native_type,
+            is_option: wrapping_type == WrappingType::Option
+                || wrapping_type == WrappingType::OptionVec,
+            is_vec: wrapping_type == WrappingType::Vec || wrapping_type == WrappingType::OptionVec,
+            is_result: false,
+            is_cow: false,
+            is_borrow: false,
+        }
+    }
+}
+
+impl TypeFFI {
+    /// Generates a `TokenStream` for turning an argument of the FFI type represented by `self` into
+    /// a native Rust type.
+    ///
+    #[must_use]
+    pub fn argument_into_rust(
+        &self,
+        field_name: &TokenStream,
+        has_custom_implementation: bool,
+    ) -> TokenStream {
+        // All FFIArrayT types have a `From<FFIArrayT> for Vec<T>` impl, so we can treat them all
+        // the same for the sake of native Rust assignment.
+        if self.is_vec {
+            return quote!(#field_name.into());
+        }
+
+        match self.native_type {
```

Comment:
nbd, but this might be easier to maintain (and follow?) if you pulled up the `self.is_option` logic; maybe like

```rust
match (self.native_type, self.is_option) => {
    (TypeIdentifier::Boxed(_), true) => {
    },
    (TypeIdentifier::Boxed(_), false) => {
    }
}
```

IDK

Reply:
Yeah, this is...hard to follow. https://github.com/agrian-inc/ffi_common/issues/16